docs/internals_overview.md

```diff
@@ -0,0 +1,382 @@
+* `/llarp/config` config loading, config parsing, config defaults, all things config related goes here.
+* `/llarp/util` general dumping ground for misc helpers and such.
+* `/llarp/dns` dns parsing library fused with assorted interfaces and their implementations for the dns subsystem.
+* `/llarp/rpc` zmq api for oxend, lokinet rpc and endpoint(exit) auth rpc.
+
+platform specific parts have been slowly isolated into their distinct locations:
+
+* `/llarp/win32` most of the windows only code, omitting some of the vpn api stuff.
+* `/llarp/android` android compat shims (`/jni` contains the android vpn api bindings for jni).
+* `/llarp/apple` as much as the apple specific stuff.
+* `/llarp/linux` right now contains linux only code like things for dbus.
+
+finally, then there is `/llarp/router` which is home to a lot of our god objects and things that are owned by said god objects.
+
+---
+
+each of the layers in lokinet are implemented in the following areas of the code:
+
+#### link layer
+
+the link layer code exists in:
+
+* `/llarp/iwp`
+* `/llarp/link`
+* `/llarp/messages`
+
+iwp is our current wire protocol, that code is in `/llarp/iwp`.
+the high level interface types exist in `/llarp/link` , sort of.
+the link layer frame parser and each link layer frame kind is in `/llarp/messages`.
+
+#### onion layer
+
+the onion layer is mainly contained in:
+
+* `/llarp/path`
+* `/llarp/exit`
+* `/llarp/service`
+
+#### routing layer
+
+the routing layer is present in:
+
+* `/llarp/routing`
+* `/llarp/exit`
+* `/llarp/handlers`
+
+### flow layer
+
+the flow layer is present in:
+
+* `/llarp/path`
+* `/llarp/handlers`
+* `/llarp/exit`
+* `/llarp/service`
+
+# problems in the current state of lokinet 
+
+there are a few issues in the high level conceptual layout of lokinet.
+
+## conceptual problems
+
+there most major issues are in the link layer and the flow layer.
+
+### Link Layer
+
+the link layer has side effects that bubble up to upper layers lokinet has no ability to be aware of.
+the congestion strategy in lokinet happens as a coincidental side effect in the implementation detail of how link layer frames are transported.
+elevated packet loss causes retransmissions which will delay traffic to the next hop. 
+this behavior is desirable but because it is done at the link layer and not the onion layer, it will affect all paths going across that node.
+this behavior should be reimplemented on an upper level, propagating link layer session info upwards would help inform how and when to do this.
+
+iwp is a homegrown protocol that nothing else speaks, it needs to be replaced.
+DTLS + SCP would be a prime choice as it is what webrtc speaks, this would make a WASM port a possiblity.
+QUIC would be great for durable transmission of messages but the capabilities of it to transmit lossy data is unclear at this time.
```

Comment:
Not unclear: https://datatracker.ietf.org/doc/html/rfc9221

Reply:
gotcha